// backendless/es/files/send.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sendFile = undefined;

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _urls = require('../urls');

var _urls2 = _interopRequireDefault(_urls);

var _request = require('../request');

var _request2 = _interopRequireDefault(_request);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var sanitizeFileName = function sanitizeFileName(fileName) {
  return encodeURIComponent(fileName).replace(/'/g, '%27').replace(/"/g, '%22');
};

var sendFile = exports.sendFile = function sendFile(options) {
  var url = _urls2.default.filePath(options.path) + '/' + sanitizeFileName(options.fileName);
  var query = {};

  if (_utils2.default.isBoolean(options.overwrite)) {
    query.overwrite = options.overwrite;
  }

  return _request2.default.post({
    url: url,
    query: query,
    form: { file: options.file },
    asyncHandler: options.asyncHandler
  });
};
