// backendless/es/cache/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _desc, _value, _obj, _init, _init2, _init3, _init4, _init5, _init6;

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _decorators = require('../decorators');

var _put = require('./put');

var _get = require('./get');

var _remove = require('./remove');

var _contains = require('./contains');

var _expireIn = require('./expire-in');

var _expireAt = require('./expire-at');

var _parsers = require('./parsers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var Cache = (_dec = (0, _decorators.deprecated)('Backendless.Cache', 'Backendless.Cache.put'), _dec2 = (0, _decorators.deprecated)('Backendless.Cache', 'Backendless.Cache.get'), _dec3 = (0, _decorators.deprecated)('Backendless.Cache', 'Backendless.Cache.remove'), _dec4 = (0, _decorators.deprecated)('Backendless.Cache', 'Backendless.Cache.contains'), _dec5 = (0, _decorators.deprecated)('Backendless.Cache', 'Backendless.Cache.expireIn'), _dec6 = (0, _decorators.deprecated)('Backendless.Cache', 'Backendless.Cache.expireAt'), (_obj = {
  putSync: _utils2.default.synchronized(_put.put),
  put: _utils2.default.promisified(_put.put),

  getSync: _utils2.default.synchronized(_get.get),
  get: _utils2.default.promisified(_get.get),

  removeSync: _utils2.default.synchronized(_remove.remove),
  remove: _utils2.default.promisified(_remove.remove),

  containsSync: _utils2.default.synchronized(_contains.contains),
  contains: _utils2.default.promisified(_contains.contains),

  expireInSync: _utils2.default.synchronized(_expireIn.expireIn),
  expireIn: _utils2.default.promisified(_expireIn.expireIn),

  expireAtSync: _utils2.default.synchronized(_expireAt.expireAt),
  expireAt: _utils2.default.promisified(_expireAt.expireAt),

  //TODO: do we need it?
  setObjectFactory: function setObjectFactory(objectName, factoryMethod) {
    _parsers.Parsers.set(objectName, factoryMethod);
  }
}, (_applyDecoratedDescriptor(_obj, 'putSync', [_dec], (_init = Object.getOwnPropertyDescriptor(_obj, 'putSync'), _init = _init ? _init.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getSync', [_dec2], (_init2 = Object.getOwnPropertyDescriptor(_obj, 'getSync'), _init2 = _init2 ? _init2.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init2;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'removeSync', [_dec3], (_init3 = Object.getOwnPropertyDescriptor(_obj, 'removeSync'), _init3 = _init3 ? _init3.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init3;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'containsSync', [_dec4], (_init4 = Object.getOwnPropertyDescriptor(_obj, 'containsSync'), _init4 = _init4 ? _init4.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init4;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'expireInSync', [_dec5], (_init5 = Object.getOwnPropertyDescriptor(_obj, 'expireInSync'), _init5 = _init5 ? _init5.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init5;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'expireAtSync', [_dec6], (_init6 = Object.getOwnPropertyDescriptor(_obj, 'expireAtSync'), _init6 = _init6 ? _init6.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init6;
  }
}), _obj)), _obj));

exports.default = Cache;
