// backendless/lib/files/count.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFileCount = getFileCount;

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _urls = require('../urls');

var _urls2 = _interopRequireDefault(_urls);

var _request = require('../request');

var _request2 = _interopRequireDefault(_request);

var _async = require('../request/async');

var _async2 = _interopRequireDefault(_async);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getFileCount(path, pattern, recursive, countDirectories, asyncHandler) {

  if (countDirectories instanceof _async2.default) {
    asyncHandler = countDirectories;
    countDirectories = undefined;
  } else if (recursive instanceof _async2.default) {
    asyncHandler = recursive;
    recursive = undefined;
    countDirectories = undefined;
  } else if (pattern instanceof _async2.default) {
    asyncHandler = pattern;
    pattern = undefined;
    recursive = undefined;
    countDirectories = undefined;
  }

  var query = {
    action: 'count',
    pattern: pattern !== undefined ? pattern : '*',
    recursive: !!recursive,
    countDirectories: !!countDirectories
  };

  if (!path || !_utils2.default.isString(path)) {
    throw new Error('Files "path" must not be empty and must be String');
  }

  if (!query.pattern || !_utils2.default.isString(query.pattern)) {
    throw new Error('Files "path" must not be empty and must be String');
  }

  return _request2.default.get({
    url: _urls2.default.filePath(path),
    query: query,
    isAsync: !!asyncHandler,
    asyncHandler: asyncHandler
  });
}
