// backendless/es/messaging/methods/register-device.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerDevice = registerDevice;

var _utils = require('../../utils');

var _utils2 = _interopRequireDefault(_utils);

var _urls = require('../../urls');

var _urls2 = _interopRequireDefault(_urls);

var _device = require('../../device');

var _device2 = _interopRequireDefault(_device);

var _request = require('../../request');

var _request2 = _interopRequireDefault(_request);

var _async = require('../../request/async');

var _async2 = _interopRequireDefault(_async);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function registerDevice(deviceToken, channels, expiration, asyncHandler) {
  var device = _device2.default.required();

  if (expiration instanceof _async2.default) {
    asyncHandler = expiration;
    expiration = undefined;
  }

  var data = {
    deviceToken: deviceToken,
    deviceId: device.uuid,
    os: device.platform,
    osVersion: device.version
  };

  if (_utils2.default.isArray(channels)) {
    data.channels = channels;
  }

  if (_utils2.default.isNumber(expiration) || expiration instanceof Date) {
    data.expiration = expiration instanceof Date ? expiration.getTime() / 1000 : expiration;
  }

  return _request2.default.post({
    url: _urls2.default.messagingRegistrations(),
    data: data,
    isAsync: !!asyncHandler,
    asyncHandler: asyncHandler
  });
}
