// backendless/es/messaging/methods/send-email.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sendEmail = sendEmail;

var _utils = require('../../utils');

var _utils2 = _interopRequireDefault(_utils);

var _urls = require('../../urls');

var _urls2 = _interopRequireDefault(_urls);

var _request = require('../../request');

var _request2 = _interopRequireDefault(_request);

var _bodyParts = require('../helpers/body-parts');

var _bodyParts2 = _interopRequireDefault(_bodyParts);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function sendEmail(subject, bodyParts, recipients, attachments /**, async */) {
  var responder = _utils2.default.extractResponder(arguments);
  var isAsync = !!responder;
  var data = {};

  if (subject && !_utils2.default.isEmpty(subject) && _utils2.default.isString(subject)) {
    data.subject = subject;
  } else {
    throw new Error('Subject is required parameter and must be a nonempty string');
  }

  if (bodyParts instanceof _bodyParts2.default && !_utils2.default.isEmpty(bodyParts)) {
    data.bodyparts = bodyParts;
  } else {
    throw new Error('Use Bodyparts as bodyParts argument, must contain at least one property');
  }

  if (recipients && _utils2.default.isArray(recipients) && !_utils2.default.isEmpty(recipients)) {
    data.to = recipients;
  } else {
    throw new Error('Recipients is required parameter, must be a nonempty array');
  }

  if (attachments) {
    if (_utils2.default.isArray(attachments)) {
      if (!_utils2.default.isEmpty(attachments)) {
        data.attachment = attachments;
      }
    }
  }

  function responseMessageStatus(res) {
    return res.status;
  }

  return _request2.default.post({
    url: _urls2.default.messagingEmail(),
    isAsync: isAsync,
    asyncHandler: _utils2.default.wrapAsync(responder, responseMessageStatus),
    data: data
  });
}
