// backendless/lib/geo/tracker-monitor/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _utils = require('../../utils');

var _utils2 = _interopRequireDefault(_utils);

var _tracker = require('./tracker');

var _tracker2 = _interopRequireDefault(_tracker);

var _fenceActions = require('./fence-actions');

var _fenceActions2 = _interopRequireDefault(_fenceActions);

var _startMonitoring = require('./start-monitoring');

var _stopMonitoring = require('./stop-monitoring');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var GeoTrackerMonitor = {
  reset: function reset() {
    _tracker2.default.reset();
  },


  runOnEnterAction: _utils2.default.promisified(_fenceActions2.default.enter),
  runOnEnterActionSync: _utils2.default.synchronized(_fenceActions2.default.enter),

  runOnStayAction: _utils2.default.promisified(_fenceActions2.default.stay),
  runOnStayActionSync: _utils2.default.synchronized(_fenceActions2.default.stay),

  runOnExitAction: _utils2.default.promisified(_fenceActions2.default.exist),
  runOnExitActionSync: _utils2.default.synchronized(_fenceActions2.default.exist),

  startGeofenceMonitoringWithInAppCallback: _utils2.default.promisified(_startMonitoring.startMonitoring),
  startGeofenceMonitoringWithInAppCallbackSync: _utils2.default.synchronized(_startMonitoring.startMonitoring),

  startGeofenceMonitoringWithRemoteCallback: _utils2.default.promisified(_startMonitoring.startMonitoring),
  startGeofenceMonitoringWithRemoteCallbackSync: _utils2.default.synchronized(_startMonitoring.startMonitoring),

  stopGeofenceMonitoring: _stopMonitoring.stopMonitoring
};

exports.default = GeoTrackerMonitor;
