// backendless/lib/commerce/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _dec, _dec2, _dec3, _desc, _value, _obj, _init, _init2, _init3;

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _decorators = require('../decorators');

var _validatePlayPurchase = require('./validate-play-purchase');

var _cancelPlaySubscription = require('./cancel-play-subscription');

var _getPlaySubscriptionStatus = require('./get-play-subscription-status');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var Commerce = (_dec = (0, _decorators.deprecated)('Backendless.Commerce', 'Backendless.Commerce.validatePlayPurchase'), _dec2 = (0, _decorators.deprecated)('Backendless.Commerce', 'Backendless.Commerce.cancelPlaySubscription'), _dec3 = (0, _decorators.deprecated)('Backendless.Commerce', 'Backendless.Commerce.getPlaySubscriptionStatus'), (_obj = {
  validatePlayPurchaseSync: _utils2.default.synchronized(_validatePlayPurchase.validatePlayPurchase),
  validatePlayPurchase: _utils2.default.promisified(_validatePlayPurchase.validatePlayPurchase),

  cancelPlaySubscriptionSync: _utils2.default.synchronized(_cancelPlaySubscription.cancelPlaySubscription),
  cancelPlaySubscription: _utils2.default.promisified(_cancelPlaySubscription.cancelPlaySubscription),

  getPlaySubscriptionStatusSync: _utils2.default.synchronized(_getPlaySubscriptionStatus.getPlaySubscriptionStatus),
  getPlaySubscriptionStatus: _utils2.default.promisified(_getPlaySubscriptionStatus.getPlaySubscriptionStatus)

}, (_applyDecoratedDescriptor(_obj, 'validatePlayPurchaseSync', [_dec], (_init = Object.getOwnPropertyDescriptor(_obj, 'validatePlayPurchaseSync'), _init = _init ? _init.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'cancelPlaySubscriptionSync', [_dec2], (_init2 = Object.getOwnPropertyDescriptor(_obj, 'cancelPlaySubscriptionSync'), _init2 = _init2 ? _init2.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init2;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getPlaySubscriptionStatusSync', [_dec3], (_init3 = Object.getOwnPropertyDescriptor(_obj, 'getPlaySubscriptionStatusSync'), _init3 = _init3 ? _init3.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init3;
  }
}), _obj)), _obj));

exports.default = Commerce;
