// backendless/es/geo/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _desc, _value, _obj, _init, _init2, _init3, _init4, _init5, _init6, _init7, _init8, _init9, _init10, _init11, _init12, _init13, _init14, _init15, _init16;

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _decorators = require('../decorators');

var _constants = require('./constants');

var _utils3 = require('./utils');

var _utils4 = _interopRequireDefault(_utils3);

var _trackerMonitor = require('./tracker-monitor');

var _trackerMonitor2 = _interopRequireDefault(_trackerMonitor);

var _cluster = require('./cluster');

var _cluster2 = _interopRequireDefault(_cluster);

var _point = require('./point');

var _point2 = _interopRequireDefault(_point);

var _query = require('./query');

var _query2 = _interopRequireDefault(_query);

var _savePoint = require('./save-point');

var _addCategory = require('./add-category');

var _deleteCategory = require('./delete-category');

var _getCategories = require('./get-categories');

var _deletePoint = require('./delete-point');

var _getPointsCount = require('./get-points-count');

var _getClusterPoints = require('./get-cluster-points');

var _loadPoints = require('./load-points');

var _loadMetadata = require('./load-metadata');

var _relativeFind = require('./relative-find');

var _loadFencePoints = require('./load-fence-points');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var Geo = (_dec = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.savePoint'), _dec2 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.find'), _dec3 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.loadMetadata'), _dec4 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.getClusterPoints'), _dec5 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.relativeFind'), _dec6 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.addCategory'), _dec7 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.getCategories'), _dec8 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.deleteCategory'), _dec9 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.deletePoint'), _dec10 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.getFencePoints'), _dec11 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.getGeopointCount'), _dec12 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.runOnEnterAction'), _dec13 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.runOnStayAction'), _dec14 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.runOnExitAction'), _dec15 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.startGeofenceMonitoringWithInAppCallback'), _dec16 = (0, _decorators.deprecated)('Backendless.Geo', 'Backendless.Geo.startGeofenceMonitoringWithRemoteCallback'), (_obj = {

  EARTH_RADIUS: _constants.EARTH_RADIUS,
  UNITS: _constants.UNITS,

  Utils: _utils4.default,

  Cluster: _cluster2.default,
  Point: _point2.default,
  Query: _query2.default,

  /** @deprecated */
  addPoint: function addPoint() /** geopoint, async */{
    return this.savePoint.apply(this, arguments);
  },

  savePointSync: _utils2.default.synchronized(_savePoint.savePoint),
  savePoint: _utils2.default.promisified(_savePoint.savePoint),

  findSync: _utils2.default.synchronized(_loadPoints.loadPoints),
  find: _utils2.default.promisified(_loadPoints.loadPoints),

  loadMetadataSync: _utils2.default.synchronized(_loadMetadata.loadMetadata),
  loadMetadata: _utils2.default.promisified(_loadMetadata.loadMetadata),

  getClusterPointsSync: _utils2.default.synchronized(_getClusterPoints.getClusterPoints),
  getClusterPoints: _utils2.default.promisified(_getClusterPoints.getClusterPoints),

  relativeFindSync: _utils2.default.synchronized(_relativeFind.relativeFind),
  relativeFind: _utils2.default.promisified(_relativeFind.relativeFind),

  addCategorySync: _utils2.default.synchronized(_addCategory.addCategory),
  addCategory: _utils2.default.promisified(_addCategory.addCategory),

  getCategoriesSync: _utils2.default.synchronized(_getCategories.getCategories),
  getCategories: _utils2.default.promisified(_getCategories.getCategories),

  deleteCategorySync: _utils2.default.synchronized(_deleteCategory.deleteCategory),
  deleteCategory: _utils2.default.promisified(_deleteCategory.deleteCategory),

  deletePointSync: _utils2.default.synchronized(_deletePoint.deletePoint),
  deletePoint: _utils2.default.promisified(_deletePoint.deletePoint),

  getFencePointsSync: _utils2.default.synchronized(_loadFencePoints.loadFencePoints),
  getFencePoints: _utils2.default.promisified(_loadFencePoints.loadFencePoints),

  getGeopointCountSync: _utils2.default.synchronized(_getPointsCount.getGeopointCount),
  getGeopointCount: _utils2.default.promisified(_getPointsCount.getGeopointCount),

  runOnEnterActionSync: _trackerMonitor2.default.runOnEnterActionSync,
  runOnEnterAction: _trackerMonitor2.default.runOnEnterAction,

  runOnStayActionSync: _trackerMonitor2.default.runOnStayActionSync,
  runOnStayAction: _trackerMonitor2.default.runOnStayAction,

  runOnExitActionSync: _trackerMonitor2.default.runOnExitActionSync,
  runOnExitAction: _trackerMonitor2.default.runOnExitAction,

  startGeofenceMonitoringWithInAppCallbackSync: _trackerMonitor2.default.startGeofenceMonitoringWithInAppCallbackSync,
  startGeofenceMonitoringWithInAppCallback: _trackerMonitor2.default.startGeofenceMonitoringWithInAppCallback,

  startGeofenceMonitoringWithRemoteCallbackSync: _trackerMonitor2.default.startGeofenceMonitoringWithRemoteCallbackSync,
  startGeofenceMonitoringWithRemoteCallback: _trackerMonitor2.default.startGeofenceMonitoringWithRemoteCallback,

  stopGeofenceMonitoring: _trackerMonitor2.default.stopGeofenceMonitoring
}, (_applyDecoratedDescriptor(_obj, 'savePointSync', [_dec], (_init = Object.getOwnPropertyDescriptor(_obj, 'savePointSync'), _init = _init ? _init.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'findSync', [_dec2], (_init2 = Object.getOwnPropertyDescriptor(_obj, 'findSync'), _init2 = _init2 ? _init2.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init2;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'loadMetadataSync', [_dec3], (_init3 = Object.getOwnPropertyDescriptor(_obj, 'loadMetadataSync'), _init3 = _init3 ? _init3.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init3;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getClusterPointsSync', [_dec4], (_init4 = Object.getOwnPropertyDescriptor(_obj, 'getClusterPointsSync'), _init4 = _init4 ? _init4.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init4;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'relativeFindSync', [_dec5], (_init5 = Object.getOwnPropertyDescriptor(_obj, 'relativeFindSync'), _init5 = _init5 ? _init5.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init5;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'addCategorySync', [_dec6], (_init6 = Object.getOwnPropertyDescriptor(_obj, 'addCategorySync'), _init6 = _init6 ? _init6.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init6;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getCategoriesSync', [_dec7], (_init7 = Object.getOwnPropertyDescriptor(_obj, 'getCategoriesSync'), _init7 = _init7 ? _init7.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init7;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'deleteCategorySync', [_dec8], (_init8 = Object.getOwnPropertyDescriptor(_obj, 'deleteCategorySync'), _init8 = _init8 ? _init8.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init8;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'deletePointSync', [_dec9], (_init9 = Object.getOwnPropertyDescriptor(_obj, 'deletePointSync'), _init9 = _init9 ? _init9.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init9;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getFencePointsSync', [_dec10], (_init10 = Object.getOwnPropertyDescriptor(_obj, 'getFencePointsSync'), _init10 = _init10 ? _init10.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init10;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getGeopointCountSync', [_dec11], (_init11 = Object.getOwnPropertyDescriptor(_obj, 'getGeopointCountSync'), _init11 = _init11 ? _init11.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init11;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'runOnEnterActionSync', [_dec12], (_init12 = Object.getOwnPropertyDescriptor(_obj, 'runOnEnterActionSync'), _init12 = _init12 ? _init12.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init12;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'runOnStayActionSync', [_dec13], (_init13 = Object.getOwnPropertyDescriptor(_obj, 'runOnStayActionSync'), _init13 = _init13 ? _init13.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init13;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'runOnExitActionSync', [_dec14], (_init14 = Object.getOwnPropertyDescriptor(_obj, 'runOnExitActionSync'), _init14 = _init14 ? _init14.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init14;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'startGeofenceMonitoringWithInAppCallbackSync', [_dec15], (_init15 = Object.getOwnPropertyDescriptor(_obj, 'startGeofenceMonitoringWithInAppCallbackSync'), _init15 = _init15 ? _init15.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init15;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'startGeofenceMonitoringWithRemoteCallbackSync', [_dec16], (_init16 = Object.getOwnPropertyDescriptor(_obj, 'startGeofenceMonitoringWithRemoteCallbackSync'), _init16 = _init16 ? _init16.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init16;
  }
}), _obj)), _obj));

exports.default = Geo;
