// backendless/es/counters/counter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _desc, _value, _obj, _init, _init2, _init3, _init4, _init5, _init6, _init7, _init8, _init9;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _decorators = require('../decorators');

var _methods = require('./methods');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Counter = function Counter(name) {
  _classCallCheck(this, Counter);

  if (!name || !_utils2.default.isString(name)) {
    throw new Error('Counter Name must be non empty String');
  }

  this.name = name;
};

exports.default = Counter;


var withCounterName = function withCounterName(method) {
  return function () {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return method.apply(undefined, [this.name].concat(args));
  };
};

//TODO: will be removed when remove sync methods
var namespaceLabel = 'Backendless.Counter.of(<CounterName>)';

_extends(Counter.prototype, (_dec = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.incrementAndGet'), _dec2 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.getAndIncrement'), _dec3 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.decrementAndGet'), _dec4 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.getAndDecrement'), _dec5 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.reset'), _dec6 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.get'), _dec7 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.addAndGet'), _dec8 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.getAndAdd'), _dec9 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.compareAndSet'), (_obj = {
  incrementAndGetSync: _utils2.default.synchronized(withCounterName(_methods.incrementAndGet)),
  incrementAndGet: _utils2.default.promisified(withCounterName(_methods.incrementAndGet)),

  getAndIncrementSync: _utils2.default.synchronized(withCounterName(_methods.getAndIncrement)),
  getAndIncrement: _utils2.default.promisified(withCounterName(_methods.getAndIncrement)),

  decrementAndGetSync: _utils2.default.synchronized(withCounterName(_methods.decrementAndGet)),
  decrementAndGet: _utils2.default.promisified(withCounterName(_methods.decrementAndGet)),

  getAndDecrementSync: _utils2.default.synchronized(withCounterName(_methods.getAndDecrement)),
  getAndDecrement: _utils2.default.promisified(withCounterName(_methods.getAndDecrement)),

  resetSync: _utils2.default.synchronized(withCounterName(_methods.reset)),
  reset: _utils2.default.promisified(withCounterName(_methods.reset)),

  getSync: _utils2.default.synchronized(withCounterName(_methods.get)),
  get: _utils2.default.promisified(withCounterName(_methods.get)),

  addAndGetSync: _utils2.default.synchronized(withCounterName(_methods.addAndGet)),
  addAndGet: _utils2.default.promisified(withCounterName(_methods.addAndGet)),

  getAndAddSync: _utils2.default.synchronized(withCounterName(_methods.getAndAdd)),
  getAndAdd: _utils2.default.promisified(withCounterName(_methods.getAndAdd)),

  compareAndSetSync: _utils2.default.synchronized(withCounterName(_methods.compareAndSet)),
  compareAndSet: _utils2.default.promisified(withCounterName(_methods.compareAndSet))

}, (_applyDecoratedDescriptor(_obj, 'incrementAndGetSync', [_dec], (_init = Object.getOwnPropertyDescriptor(_obj, 'incrementAndGetSync'), _init = _init ? _init.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getAndIncrementSync', [_dec2], (_init2 = Object.getOwnPropertyDescriptor(_obj, 'getAndIncrementSync'), _init2 = _init2 ? _init2.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init2;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'decrementAndGetSync', [_dec3], (_init3 = Object.getOwnPropertyDescriptor(_obj, 'decrementAndGetSync'), _init3 = _init3 ? _init3.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init3;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getAndDecrementSync', [_dec4], (_init4 = Object.getOwnPropertyDescriptor(_obj, 'getAndDecrementSync'), _init4 = _init4 ? _init4.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init4;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'resetSync', [_dec5], (_init5 = Object.getOwnPropertyDescriptor(_obj, 'resetSync'), _init5 = _init5 ? _init5.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init5;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getSync', [_dec6], (_init6 = Object.getOwnPropertyDescriptor(_obj, 'getSync'), _init6 = _init6 ? _init6.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init6;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'addAndGetSync', [_dec7], (_init7 = Object.getOwnPropertyDescriptor(_obj, 'addAndGetSync'), _init7 = _init7 ? _init7.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init7;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getAndAddSync', [_dec8], (_init8 = Object.getOwnPropertyDescriptor(_obj, 'getAndAddSync'), _init8 = _init8 ? _init8.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init8;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'compareAndSetSync', [_dec9], (_init9 = Object.getOwnPropertyDescriptor(_obj, 'compareAndSetSync'), _init9 = _init9 ? _init9.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init9;
  }
}), _obj)), _obj)));
