// backendless/es/users/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _desc, _value, _obj, _init, _init2, _init3, _init4, _init5, _init6, _init7, _init8, _init9, _init10, _init11, _init12, _init13, _init14, _init15;

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _decorators = require('../decorators');

var _register = require('./register');

var _roles = require('./roles');

var _login = require('./login');

var _logout = require('./logout');

var _update = require('./update');

var _describeClass = require('./describe-class');

var _restorePassword = require('./restore-password');

var _currentUser = require('./current-user');

var _emailConfirmation = require('./email-confirmation');

var _social = require('./social');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var Users = (_dec = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.register'), _dec2 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.getUserRoles'), _dec3 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.assignRole'), _dec4 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.unassignRole'), _dec5 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.login'), _dec6 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.describeUserClass'), _dec7 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.restorePassword'), _dec8 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.logout'), _dec9 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.getCurrentUser'), _dec10 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.update'), _dec11 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.loginWithFacebook'), _dec12 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.loginWithGooglePlus'), _dec13 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.loginWithTwitter'), _dec14 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.isValidLogin'), _dec15 = (0, _decorators.deprecated)('Backendless.Users', 'Backendless.Users.resendEmailConfirmation'), (_obj = {
  registerSync: _utils2.default.synchronized(_register.register),
  register: _utils2.default.promisified(_register.register),

  getUserRolesSync: _utils2.default.synchronized(_roles.getUserRoles),
  getUserRoles: _utils2.default.promisified(_roles.getUserRoles),

  assignRoleSync: _utils2.default.synchronized(_roles.assignRole),
  assignRole: _utils2.default.promisified(_roles.assignRole),

  unassignRoleSync: _utils2.default.synchronized(_roles.unassignRole),
  unassignRole: _utils2.default.promisified(_roles.unassignRole),

  loginSync: _utils2.default.synchronized(_login.login),
  login: _utils2.default.promisified(_login.login),

  describeUserClassSync: _utils2.default.synchronized(_describeClass.describeUserClass),
  describeUserClass: _utils2.default.promisified(_describeClass.describeUserClass),

  restorePasswordSync: _utils2.default.synchronized(_restorePassword.restorePassword),
  restorePassword: _utils2.default.promisified(_restorePassword.restorePassword),

  logoutSync: _utils2.default.synchronized(_logout.logout),
  logout: _utils2.default.promisified(_logout.logout),

  getCurrentUserSync: _utils2.default.synchronized(_currentUser.getCurrentUser),
  getCurrentUser: _utils2.default.promisified(_currentUser.getCurrentUser),

  updateSync: _utils2.default.synchronized(_update.update),
  update: _utils2.default.promisified(_update.update),

  loginWithFacebookSync: _utils2.default.synchronized(_social.loginWithFacebook),
  loginWithFacebook: _utils2.default.promisified(_social.loginWithFacebook),
  loginWithFacebookSdk: _social.loginWithFacebookSdk,

  loginWithGooglePlusSync: _utils2.default.synchronized(_social.loginWithGooglePlus),
  loginWithGooglePlus: _utils2.default.promisified(_social.loginWithGooglePlus),
  loginWithGooglePlusSdk: _social.loginWithGooglePlusSdk,

  loginWithTwitterSync: _utils2.default.synchronized(_social.loginWithTwitter),
  loginWithTwitter: _utils2.default.promisified(_social.loginWithTwitter),

  isValidLoginSync: _utils2.default.synchronized(_currentUser.isValidLogin),
  isValidLogin: _utils2.default.promisified(_currentUser.isValidLogin),

  resendEmailConfirmationSync: _utils2.default.synchronized(_emailConfirmation.resendEmailConfirmation),
  resendEmailConfirmation: _utils2.default.promisified(_emailConfirmation.resendEmailConfirmation),

  loggedInUser: _currentUser.loggedInUser

}, (_applyDecoratedDescriptor(_obj, 'registerSync', [_dec], (_init = Object.getOwnPropertyDescriptor(_obj, 'registerSync'), _init = _init ? _init.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getUserRolesSync', [_dec2], (_init2 = Object.getOwnPropertyDescriptor(_obj, 'getUserRolesSync'), _init2 = _init2 ? _init2.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init2;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'assignRoleSync', [_dec3], (_init3 = Object.getOwnPropertyDescriptor(_obj, 'assignRoleSync'), _init3 = _init3 ? _init3.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init3;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'unassignRoleSync', [_dec4], (_init4 = Object.getOwnPropertyDescriptor(_obj, 'unassignRoleSync'), _init4 = _init4 ? _init4.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init4;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'loginSync', [_dec5], (_init5 = Object.getOwnPropertyDescriptor(_obj, 'loginSync'), _init5 = _init5 ? _init5.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init5;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'describeUserClassSync', [_dec6], (_init6 = Object.getOwnPropertyDescriptor(_obj, 'describeUserClassSync'), _init6 = _init6 ? _init6.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init6;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'restorePasswordSync', [_dec7], (_init7 = Object.getOwnPropertyDescriptor(_obj, 'restorePasswordSync'), _init7 = _init7 ? _init7.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init7;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'logoutSync', [_dec8], (_init8 = Object.getOwnPropertyDescriptor(_obj, 'logoutSync'), _init8 = _init8 ? _init8.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init8;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getCurrentUserSync', [_dec9], (_init9 = Object.getOwnPropertyDescriptor(_obj, 'getCurrentUserSync'), _init9 = _init9 ? _init9.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init9;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'updateSync', [_dec10], (_init10 = Object.getOwnPropertyDescriptor(_obj, 'updateSync'), _init10 = _init10 ? _init10.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init10;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'loginWithFacebookSync', [_dec11], (_init11 = Object.getOwnPropertyDescriptor(_obj, 'loginWithFacebookSync'), _init11 = _init11 ? _init11.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init11;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'loginWithGooglePlusSync', [_dec12], (_init12 = Object.getOwnPropertyDescriptor(_obj, 'loginWithGooglePlusSync'), _init12 = _init12 ? _init12.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init12;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'loginWithTwitterSync', [_dec13], (_init13 = Object.getOwnPropertyDescriptor(_obj, 'loginWithTwitterSync'), _init13 = _init13 ? _init13.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init13;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'isValidLoginSync', [_dec14], (_init14 = Object.getOwnPropertyDescriptor(_obj, 'isValidLoginSync'), _init14 = _init14 ? _init14.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init14;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'resendEmailConfirmationSync', [_dec15], (_init15 = Object.getOwnPropertyDescriptor(_obj, 'resendEmailConfirmationSync'), _init15 = _init15 ? _init15.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init15;
  }
}), _obj)), _obj));

exports.default = Users;
